docs/modules/sql/pages/user-defined-types.adoc

```diff
@@ -1,173 +1,161 @@
-If a Type hierarchy contains cycles, any mapping using any of these types (provided that type is not itself an acyclic branch)
-will have `INSERT` and `UPDATE` commands disabled.
-Additionally, support for cycles also means no validation for existence of custom types at the time of `CREATE TYPE` execution.
-**Type hierarchies are only verified for consistency upon actual use in `CREATE MAPPING`.**
+CREATE [OR REPLACE] TYPE [IF NOT EXISTS] MyTypeName [(
+    colName colType,
+    ...
+)] [OPTIONS (
+    'javaClass'='com.mypackage.MyJavaClass' <1>
+  | 'compactTypeName'='MyCompactRecordName' <2>
+  | 'portableFactoryId'='123', 'portableClassId'='456', ['portableVersion'='789'] <3>
+  | 'avroSchema'='{"type":"record","name":"myType","fields":[{"name":"colName","type":"colType"},...]}' <4>
+)]
+----
+<1> In `java` format, you can use the `javaClass` option to override the type class with a subclass. This is necessary if the original type class is abstract or interface, and you want to use `INSERT` and `UPDATE` statements.
+<2> In `portable` format, you can use the `portableFactoryId`, `portableClassId` and `portableVersion` options to specify a portable ID for the type. This is only effective when the type is used for `__key` or `this` fields. `portableVersion` defaults to 0 if not specified. Portable ID components must be defined together; incomplete definitions have the same effect as no definition.
+<3> In `compact` format, you can use the `compactTypeName` option to specify the Compact record name. If unspecified, it defaults to `<TypeName>CompactType`.
+<4> In `avro` format, you can use the `avroSchema` option to specify an inline Avro schema. This is only effective when the type is used for `__key` or `this` fields.
+
+[NOTE]
+====
+. `EXTERNAL NAME` aliases are not supported for UDTs; column names must have the same name as their corresponding Java class field or Portable/Compact/Avro schema field.
+. You can mix options that belong to different formats. When you create a mapping that references your UDT, the relevant options are used and the others are ignored, which makes it possible to use a UDT in multiple mappings having different formats.
+====
 
 === Replacing Types and Type Consistency
 Currently, there is a limitation on the replacement of existing types:
-if the replaced type was already used in a mapping, you need to fully replace that mapping
+if the replaced type was already used in a mapping, you need to recreate that mapping
 to update its data type information using the `DROP MAPPING` and `CREATE MAPPING` statements.
 However, if the type hierarchy was not used in a mapping, any type in that hierarchy can be safely
 changed, and these changes will appear in the new mapping. This is because the links
-between types are symbolic (based on the name only), and they're only "materialized" once used in a mapping.
+between types are symbolic (based on the name only), and they are materialized only when used in a mapping.
```

Comment:
and fields for fieldless UDT are captured and stored at this time

Reply:
In Feature Overview, it says
> Type fields are optional and if not specified, they are resolved from the class/schema when the type is used in a `CREATE MAPPING` statement.

Thus, both the field resolution and materialization happens when the type is used in mapping. They are mentioned in different sections, which is, I think, not a shortcoming since the documentation is not pivoted around mapping/type lifecycle.

Reply:
let's move discussion to https://github.com/hazelcast/hz-docs/pull/969#discussion_r1466279286